packages/shared/common/src/api/subsystem/DataSystem/CompositeDataSource.ts

```diff
@@ -0,0 +1,291 @@
+ * Given a {@link DataSourceState}, how long to wait before transitioning.
+ */
+export type TransitionConditions = {
+  [k in DataSourceState]?: { durationMS: number; transition: Transition };
+};
+
+interface TransitionRequest {
+  transition: Transition;
+  err?: Error;
+}
+
+/**
+ * The {@link CompositeDataSource} can combine a number of {@link DataSystemInitializer}s and {@link DataSystemSynchronizer}s
+ * into a single {@link DataSource}, implementing fallback and recovery logic internally to choose where data is sourced from.
+ */
+export class CompositeDataSource implements DataSource {
+  // TODO: SDK-856 async notification if initializer takes too long
+  // TODO: SDK-1044 utilize selector from initializers
+  private readonly _defaultFallbackTimeMs = 2 * 60 * 1000;
+  private readonly _defaultRecoveryTimeMs = 5 * 60 * 1000;
+
+  private _initPhaseActive: boolean = true;
+  private _currentPosition: number = 0;
+
+  private _stopped: boolean = true;
+  private _externalTransitionPromise: Promise<TransitionRequest>;
+  private _externalTransitionResolve?: (value: TransitionRequest) => void;
+
+  /**
+   * @param _initializers factories to create {@link DataSystemInitializer}s, in priority order.
+   * @param _synchronizers factories to create  {@link DataSystemSynchronizer}s, in priority order.
+   */
+  constructor(
+    private readonly _initializers: InitializerFactory[],
+    private readonly _synchronizers: SynchronizerFactory[],
+    private readonly _transitionConditions: TransitionConditions,
+    private readonly _backoff: Backoff,
+  ) {
+    this._externalTransitionPromise = new Promise<TransitionRequest>((tr) => {
+      this._externalTransitionResolve = tr;
+    });
+    this._initPhaseActive = true;
+    this._currentPosition = 0;
+  }
+
+  async run(
+    dataCallback: (basis: boolean, data: Data) => void,
+    statusCallback: (status: DataSourceState, err?: any) => void,
+  ): Promise<void> {
+    if (!this._stopped) {
+      // don't allow multiple simultaneous runs
+      return;
+    }
+    this._stopped = false;
+
+    let lastTransition: Transition = Transition.None;
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      const currentDS: DataSource | undefined = this._pickDataSource(lastTransition);
+      const internalTransitionPromise = new Promise<TransitionRequest>((transitionResolve) => {
+        if (currentDS) {
+          // these local variables are used for handling automatic transition related to data source status (ex: recovering to primary after
+          // secondary has been valid for N many seconds)
+          let lastState: DataSourceState | undefined;
+          let cancelScheduledTransition: (() => void) | undefined;
+
+          // this callback handler can be disabled and ensures only one transition request occurs
+          const callbackHandler = new CallbackHandler(
+            (basis: boolean, data: Data) => {
+              this._backoff.success(Date.now());
```

Comment:
Timestamp is only meant to be explicitly passed for testing. 
```suggestion
              this._backoff.success();
```

Reply:
When I abstracted out the Backoff interface, I left the parameters to maintain the same level of testability and control.  Perhaps the interface shouldn't have the parameter and the interface imposes that current time is used?  I didn't think super far into how that might evolve over time.  Thinking...